Implement second core on LPC54114

stanislav-poboril opened the pull request:
Introduces support for the Cortex-M0+ core of LPC54114 in addition to the existing support for the Cortex-M4  core. Introduces support for mailbox communication between the two cores. This PR replaces #5960, which has been abandoned due to RPMsg-Lite solution not chosen as the standard AMP solution in Zephyr. So it contains the same commits (except for RPMsg-Lite related ones) and it's also rebased to the recent code in master branch.

The new mailbox (ipm_mcux) driver could be possibly used by the chosen AMP solution - OpenAMP - to enable assymetric multiprocessing.

* Adds support for the slave core (Cortex M0+) on LPC54114
* Adds an IPM shim driver using the mcux mailbox driver
* Adds IPM samples

stanislav-poboril committed:
ext: Add dual core startup code for lpc54114 based on mcux 2.2.0

The new file startup_LPC54114_cm4.S is a copy of the file
devices/LPC54114/gcc/startup_LPC54114_cm4.S from mcux 2.2.0.
It contains platform specific initialization code for both cores.

Origin: NXP MCUXpresso SDK 2.2.0
URL: mcux.nxp.com
Maintained-by: External

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
lpc: Added support for slave Cortext-M0+ core of lpc54114 soc

Add soc configuration support and dts files for nxp_lpc54xxx_m0.

Adjusted nxp_lpc54xxx soc linker, configuration and dts files
for the presence of slave core.

Added startup hook support in ARM Cortext-M arch to allow
the initialization of slave core from startup and reset.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
mcux: Add MCUX IPM driver for lpc and kinetis socs

Add driver for MCUX mailbox which can be used for lpcxpresso54114
and other lpc and kinetis socs.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
lpc: Add code to include slave core binary for lpc54114 from mcux 2.2.0

Add ext code to be able to include image of slave core in primary
core image.

Origin: NXP MCUXpresso SDK 2.2.0
URL: mcux.nxp.com
Maintained-by: External

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
sample: Add MCUX IPM sample application

Add MCUX IPM sample application. It can be run on lpcxpresso54114
board at the moment.

First, the slave core image has to be built from
samples/subsys/ipc/ipm_mcux/remote/.
Then, buid primary core image from samples/subsys/ipc/ipm_mcux/master/.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
lpc: Add support of slave core in lpcxpresso54114 board

Add lpcxpresso54114_m0 board dts, yaml, fixup, docs, configuration,
pinmux and header files.

The primary core changed to use sram0 instead of sram1 now.
The secondary core uses sram1 for its code and sram2 as a sram.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

galak reviewed (commented):
startup_LPC54114_cm4.S doesn't seem to exist in SDK MCUx 2.2.1

dbkinder commented on boards/arm/lpcxpresso54114_m0/doc/lpcxpresso54114.rst:
```diff
@@ -0,0 +1,155 @@
+``Firmware_JLink_LPC-Link2_20160923.bin``. Serial communication problems, such
+as dropping characters, have been observed with older versions of the firmware.
+
+The code for the secondary core is linked into primary core binary file.
```
``into the primary``

stanislav-poboril replied:
Ok, fixed.

dbkinder commented on boards/arm/lpcxpresso54114_m0/doc/lpcxpresso54114.rst:
```diff
@@ -0,0 +1,155 @@
+as dropping characters, have been observed with older versions of the firmware.
+
+The code for the secondary core is linked into primary core binary file.
+Startup code copies secondary core's code into appropriate location in RAM
```
``copies the secondary core's code into an appropriate``

stanislav-poboril replied:
Ok, fixed.

galak commented on boards/arm/lpcxpresso54114_m0/pinmux.c:
```diff
@@ -0,0 +1,94 @@
+#include <fsl_common.h>
+#include <fsl_lpc_iocon.h>
+
+#define IOCON_PIO_DIGITAL_EN	0x80u
```
these defines should be somewhere common.

stanislav-poboril replied:
Ok, fixed. This was not in MCUxpresso SDK device header file, but I have added the file pin_mux.h from SDK to ext. It defines these things.

galak commented on boards/arm/lpcxpresso54114_m0/doc/lpcxpresso54114.rst:
```diff
@@ -0,0 +1,155 @@
+
+   ***** BOOTING ZEPHYR OS v1.10.99 - BUILD: Feb  7 2018 20:32:27 *****
+   Hello World from MASTER! arm
+   Received: 1
```
The Recieved:1 ... should probalby be removed from here.

stanislav-poboril replied:
I have replaced by

Received: 1
...
Received: 99

to not include all lines.

galak commented on boards/arm/lpcxpresso54114_m0/lpcxpresso54114_m0.yaml:
```diff
@@ -0,0 +1,15 @@
+type: mcu
+arch: arm
+ram: 64
+flash: 256
```
should this be 64 since its pointing to sram1?

stanislav-poboril replied:
Ok, fixed.

galak commented on boards/arm/lpcxpresso54114_m0/lpcxpresso54114_m0.yaml:
```diff
@@ -0,0 +1,15 @@
+name: NXP LPCXpresso54114
+type: mcu
+arch: arm
+ram: 64
```
Should this be 32 since its pointing at sram2?

stanislav-poboril replied:
Ok, fixed.

galak commented on boards/arm/lpcxpresso54114_m0/board.h:
```diff
@@ -0,0 +1,36 @@
+#include <soc.h>
+
+/* Red LED */
+#define RED_GPIO_NAME	CONFIG_GPIO_MCUX_LPC_PORT0_NAME
```
Would be nice if this moved to being in dts.

stanislav-poboril replied:
Blinky demo no longer builds even for master core on master branch.
I have put moved it into DTS but commented it there. Need to add missing gpio sections in soc DTS.
I have tried it but hasn't been able to make it work.
So blinky demo still does not work.

stanislav-poboril replied:
leds are fixed and in dts now

galak commented on boards/arm/lpcxpresso54114/lpcxpresso54114.dts:
```diff
@@ -17,7 +17,7 @@
 
 	chosen {
-		zephyr,sram = &sram1;
+		zephyr,sram = &sram0;
```
add a comment in commit message about this change.

stanislav-poboril replied:
Ok, modified commit message.

galak commented on arch/arm/soc/nxp_lpc/lpc54xxx/dts.fixup:
```diff
@@ -6,8 +6,6 @@
 
 /* SoC level DTS fixup file */
 
-#define CONFIG_NUM_IRQ_PRIO_BITS		ARM_V7M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
```
Do this as:
```
#if defined(CONFIG_SOC_LPC54114_M0)
#define CONFIG_NUM_IRQ_PRIO_BITS                ARM_V6M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
#else 
#define CONFIG_NUM_IRQ_PRIO_BITS                ARM_V7M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
#endif
```

stanislav-poboril replied:
Ok, changed.

galak commented on ext/hal/nxp/mcux/devices/LPC54114/incbin.S:
```diff
@@ -0,0 +1,44 @@
+; * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+; * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+; * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+; */
```
Not really sure the value of having this imported from the SDK.  Can you guys just do a version that is the arch/soc code that is Apache licensed.

There isn't really anything unique here.

stanislav-poboril replied:
This is not changed yet. The file content is copied from SDK and the license text says it has to stay there. So would have to rewrite it with different implementation actually. Otherwise our license scans on SDK would probably detect this match with Zephyr.
Why is it necessary to not have it in ext and have it in soc with Apache license?

galak replied:
Well for one, we probably want a solution that is not just NXP specific.  We have other SoCs that will probably support multiple cores and we want something that works for all of them.

stanislav-poboril replied:
Changed commit message as the file is actually from SDK 2.2.0, without Clear BSD license.

galak reviewed (changes requested):
Various areas that need cleaning up.

stanislav-poboril commented:
> startup_LPC54114_cm4.S doesn't seem to exist in SDK MCUx 2.2.1

I don't have a 2.2.1 SDK package, but I have downloaded 2.3.1 and it is there in devices/LPC54114/gcc/startup_LPC54114_cm4.S. The file amp_startup.S leaves stuff not necessary in Zephyr out, but the rest is based on the SDK file. I will change the commit message to mention mcux 2.3.1 instead of 2.2.1.

galak commented:
> I don't have a 2.2.1 SDK package, but I have downloaded 2.3.1 and it is there in devices/LPC54114/gcc/startup_LPC54114_cm4.S. The file amp_startup.S leaves stuff not necessary in Zephyr out, but the rest is based on the SDK file. I will change the commit message to mention mcux 2.3.1 instead of 2.2.1.

I think it should be 2 commits, include saying 2.3.1.  So first import the file exactly as is from the SDK, than modify it in a separate commit for what we need in Zephyr.

stanislav-poboril committed:
ext: Modified lpc54114 startup code from mcux for use with Zephyr.

The instructions performed by standard Zephyr startup files are removed
from the file startup_LPC54114_cm4.S. Introduced the section
_PlatformInit which will be called when platform specific initialization
is needed.

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril commented on boards/arm/lpcxpresso54114_m0/doc/lpcxpresso54114.rst:
```diff
@@ -0,0 +1,155 @@
+Segger J-Link firmware variant.
+
+Before you start using Zephyr on the LPCXpresso54114, download and run
+`LPCScrypt`_ to update the LPC-Link2 firmware to the latest version, currently
```
Ok, should be fixed now.

stanislav-poboril commented on boards/arm/lpcxpresso54114_m0/doc/lpcxpresso54114.rst:
```diff
@@ -0,0 +1,155 @@
+=========
+
+You can debug an application in the usual way. Here is an example for the
+:ref:`ipm_mcux` application.
```
Ok, fixed.

stanislav-poboril commented:
> I think it should be 2 commits, include saying 2.3.1. So first import the file exactly as is from the SDK, than modify it in a separate commit for what we need in Zephyr.

Separated into two commits.

dbkinder commented on samples/subsys/ipc/ipm_mcux/README.rst:
```diff
@@ -0,0 +1,138 @@
+Overview
+********
+
+This application will use mailbox to send messages from one processor core
```
How about:

    The :ref:`lpcxpresso54114` board has two core processors (M4F and dual M0).
    This sample application uses a mailbox to send messages from one processor core
    to the other.

stanislav-poboril replied:
Changed.

dbkinder commented on samples/subsys/ipc/ipm_mcux/README.rst:
```diff
@@ -0,0 +1,138 @@
+   ***** Booting Zephyr OS v1.11.0-764-g4e3007a *****
+   Hello World from MASTER! arm
+   Received: 1
+   Received: 2
```
Let's do the same here with:

    Received 1
    ...
    Received 99

stanislav-poboril replied:
Changed.

dbkinder commented on samples/subsys/ipc/ipm_mcux/README.rst:
```diff
@@ -0,0 +1,138 @@
+.. _ipm-mcux-sample:
```
Since this is a new documentation area in the samples, you'll need to create an index/TOC page named ``samples/subsys/ipc/ipc.rst`` with the contents:

    .. _ipc_samples:

    IPC Samples
    ###########

    .. toctree::
       :maxdepth: 1
       :glob:

       **/*

(This will fix the "not included in any TOC" error.)

stanislav-poboril replied:
Added. Is there a way to validate the rst docs myself before pushing? Thanks.

stanislav-poboril replied:
Thanks.

stanislav-poboril commented:
Thank you for your comments. Everything should fixed now (except for the external file incbin.S which will stay external). Could you review once more and approve or have further comments? Thanks.

galak commented on arch/arm/soc/nxp_lpc/lpc54xxx/Kconfig.defconfig.lpc54114_m0:
```diff
@@ -0,0 +1,42 @@
+	string
+	default lpc54114_m0
+
+config CPU_CORTEX_M_HAS_VTOR
```
this should be a select under config SOC_LPC54114_M0

stanislav-poboril replied:
Done.

galak commented on arch/arm/soc/nxp_lpc/lpc54xxx/soc.h:
```diff
@@ -23,6 +23,17 @@ extern "C" {
 #include <misc/util.h>
 #include <fsl_common.h>
+
+/* Address of RAM, where the image for core1 should be copied */
```
The addresses seem like the should be board specific.

stanislav-poboril replied:
The address should be shared by two boards in fact: the M4 version and M0 version. But these are separate boards in Zephyr now, so unfortunately there will stay some sort of duplication. But for sure it can be done at least better than now.

galak commented on ext/hal/nxp/mcux/CMakeLists.txt:
```diff
@@ -5,7 +5,12 @@ string(TOUPPER ${CONFIG_SOC} MCUX_DEVICE)
 else()
-  set(MCUX_CPU CPU_${CONFIG_SOC_PART_NUMBER})
+  if("${MCUX_DEVICE}" STREQUAL "LPC54114_M0")
+  set(MCUX_CPU CPU_${CONFIG_SOC_PART_NUMBER}_cm0plus)
```
nit: should ident these for if block

stanislav-poboril replied:
Fixed.

galak commented on arch/arm/soc/nxp_lpc/lpc54xxx/soc.c:
```diff
@@ -35,6 +35,8 @@
 static ALWAYS_INLINE void clkInit(void)
 {
+
+#ifdef CONFIG_SOC_LPC54114
```
Wonder if this would be a bit more clear as #ifndef CONFIG_SOC_LPC54114_M0, or do we have a CONFIG_SOC_LPC54114_M4 define?

Maybe a comment, about doing this only on the master/m4

MaureenHelm replied:
Do we want to change the name of the original soc to append `_M4` so it's explicit?

galak replied:
I think renaming to _M4 would be good.

stanislav-poboril replied:
Done rename to _M4.

Note that SOCs are _M4 and _M0 now, while board has no suffix for M4 configuration and board with M0 configuration has suffix for M0. This is because the name of the board already existed without _M4 suffix before.

galak reviewed (changes requested):
Some additional review comments/cleanups.

galak commented on ext/hal/nxp/mcux/CMakeLists.txt:
```diff
@@ -30,3 +35,7 @@ add_subdirectory_ifdef(
+
+enable_language(C ASM)
+
+zephyr_sources_ifdef(CONFIG_SOC_LPC54114 devices/${MCUX_DEVICE}/amp_startup.S)
```
Any reason not to keep the same path & filename as the SDK?

stanislav-poboril replied:
Fixed, kept the same name.

MaureenHelm commented on arch/arm/soc/nxp_lpc/lpc54xxx/Kconfig.defconfig.lpc54114:
```diff
@@ -32,4 +32,12 @@ config USART_MCUX_LPC
 
 endif # SERIAL
 
+config SLAVE_CORE_MCUX
```
This needs to move to Kconfig.soc

MaureenHelm commented on arch/arm/soc/nxp_lpc/lpc54xxx/linker.ld:
```diff
@@ -13,4 +13,23 @@
+/* Specify the memory areas */
+MEMORY
+{
+  m_core1_image         (RX)  : ORIGIN = 0x00030000, LENGTH = 0x00010000
```
Can these values come from dts?

MaureenHelm commented on dts/arm/nxp/nxp_lpc54xxx.dtsi:
```diff
@@ -39,7 +30,10 @@
+			/* reduced from 0x40000 to 0x30000
+			 * to leave room for secondary core
+			 */
+			reg = <0 0x30000>;
```
Would it be better to use partitions instead, like we do for mcuboot?

stanislav-poboril replied:
The partition names currently available for mcuboot might be a little misleading for this case, but I can try to change it.

ioannisg commented on arch/arm/core/cortex_m/Kconfig:
```diff
@@ -285,6 +285,12 @@ config SW_VECTOR_RELAY
+config PLATFORM_SPECIFIC_INIT
+	bool
+	prompt "Enable platform (SOC) specific startup hook"
+	default n
```
Some help text, here, would be very nice, since this will be a generic Cortex-M K-option (even though the option is self-descriptive)

stanislav-poboril replied:
Added a simple help text.

ioannisg commented on arch/arm/soc/nxp_lpc/lpc54xxx/Kconfig.series:
```diff
@@ -7,7 +7,8 @@
 config SOC_SERIES_LPC54XXX
 	bool "LPC LPC54xxx Series MCU"
-	select CPU_CORTEX_M4
+	select CPU_CORTEX_M
```
Aren't CPU_CORTEX_M explicit selections obsolete, now @galak ?

ioannisg commented on drivers/ipm/Kconfig:
```diff
@@ -22,3 +22,10 @@ config IPM_QUARK_SE_MASTER
+	default n
+	depends on IPM && HAS_MCUX
+	help
+	Driver for MCUX mailbox
```
Fix indentation for help message

dbkinder reviewed (approved):
Doc changes look OK. Thanks.

stanislav-poboril commented:
This PR is replaced by #7558 (already merged) and #7560.
